casanovo/data/db_utils.py

```diff
@@ -0,0 +1,389 @@
+        Digests a FASTA file and returns the peptides, their masses, and associated protein.
+
+        Parameters
+        ----------
+        fasta_filename : str
+            Path to the FASTA file.
+        enzyme : str
+            The enzyme to use for digestion.
+            See pyteomics.parser.expasy_rules for valid enzymes.
+            Can also be a regex pattern.
+        digestion : str
+            The type of digestion to perform. Either 'full', 'partial' or 'non-specific'.
+        missed_cleavages : int
+            The number of missed cleavages to allow.
+        max_mods : int
+            The maximum number of modifications to allow per peptide.
+        min_peptide_length : int
+            The minimum length of peptides to consider.
+        max_peptide_length : int
+            The maximum length of peptides to consider.
+
+        Returns
+        -------
+        pep_table : pd.DataFrame
+            A Pandas DataFrame with peptide, mass,
+            and protein columns. Sorted by neutral mass in ascending order.
+        prot_map : dict
+            A dictionary mapping peptides to associated proteins.
+        """
+        # Verify the existence of the file:
+        if not os.path.isfile(fasta_filename):
+            logger.error("File %s does not exist.", fasta_filename)
+            raise FileNotFoundError(f"File {fasta_filename} does not exist.")
+
+        peptide_list = []
+        if digestion not in ["full", "partial", "non-specific"]:
+            logger.error("Digestion type %s not recognized.", digestion)
+            raise ValueError(f"Digestion type {digestion} not recognized.")
+        if enzyme not in parser.expasy_rules:
+            logger.info(
+                "Enzyme %s not recognized. Interpreting as cleavage rule.",
+                enzyme,
+            )
+        valid_aa = set(list(self.residues.keys()) + ["C"])
+        if digestion == "non-specific":
+            for header, seq in fasta.read(fasta_filename):
+                pep_set = []
+                # Generate all possible peptides
+                for i in range(len(seq)):
+                    for j in range(i + 1, len(seq) + 1):
+                        pep_set.append(seq[i:j])
+                protein = header.split()[0]
+                for pep in pep_set:
+                    if (
+                        len(pep) >= min_peptide_length
+                        and len(pep) <= max_peptide_length
+                    ):
+                        if any(aa not in valid_aa for aa in pep):
+                            logger.warn(
+                                "Skipping peptide with unknown amino acids: %s",
+                                pep,
+                            )
+                        else:
+                            peptide_list.append((pep, protein))
+        else:
+            semi = digestion == "partial"
+            for header, seq in fasta.read(fasta_filename):
+                pep_set = parser.cleave(
```

Comment:
**todo:** This is pretty much code duplication with the non-specific setting above. Instead, generalize it to:
1. You have a function to go from proteins to peptides, which is either the non-specific implementation or `parser.cleave`. This function returns a peptide generator.
2. Then iterate over all of the peptides from that generator and process them.

That way, the checking and adding to the peptide-to-protein dict is implemented only once.

Reply:
Moved repeated code to method